### ISSUE TITLE
Api version `2024-04-02` for resources `Resource`

### ISSUE THREAD

**Issue by ruben-janssens** [pulumi/pulumi-azure-native#3524]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

API version 2024-04-02 is needed to create a `Resource` with type `GitHub.Network/networkSettings`. Currently no API versions support this resource type.

Error message from Azure:
```
No registered resource provider found for location 'germanywestcentral' and API version '2022-09-01' for type 'networkSettings'. The supported api-versions are '2024-04-01-preview, 2023-03-15-alpha, 2023-03-15-beta, 2024-04-02, 2023-11-01-preview, 2024-04-01, 2023-11-01-rc'.
```
None of these API versions are implemented currently

**Comment by thomas11** [comment]
Thank you for reporting this, @ruben-janssens. This is an interesting API in that there are two API versions in play: the one to talk to the Microsoft.Resources provider, and the one used to create the actual resource in the target provider `GitHub.Network`.

For the Microsoft.Resources provider, 2022-09-01 is correct as defined in the Azure spec. But it's not correct for GitHub.Network. It seems that the Microsoft.Resources provider simply passes on the API version it's called under to the target provider? This could only work if Microsoft.Resources accepted any API version, though.

I wasn't able to find decisive docs or examples about this API version handling, either.

@danielrbradley I think #2467 might help with this one?

**Comment by ruben-janssens** [comment]
As a side note, this is the command used in the GitHub documentation. (We want to use this resource to create private networking in Azure)
```
. az resource create --resource-group $RESOURCE_GROUP_NAME  --name $NETWORK_SETTINGS_RESOURCE_NAME --resource-type GitHub.Network/networkSettings --properties "{ \"location\": \"$AZURE_LOCATION\", \"properties\" : {  \"subnetId\": \"/subscriptions/$SUBSCRIPTION_ID/resourceGroups/$RESOURCE_GROUP_NAME/providers/Microsoft.Network/virtualNetworks/$VNET_NAME/subnets/$SUBNET_NAME\", \"businessId\": \"$DATABASE_ID\" }}" --is-full-object --output table --query "{GitHubId:tags.GitHubId, name:name}" --api-version 2024-04-02
```
Here a version is passed with `--api-version`. This is an optional parameter I assume for the second API then?

**Comment by thomas11** [comment]
> This is an optional parameter I assume for the second API then?

That's my understanding as well, although the docs are vague.

**Comment by PhilBroderickCrezco** [comment]
Bumping as we're also experiencing this issue when trying to configure GitHub with Azure private networking. @thomas11 is this blocked by https://github.com/pulumi/pulumi-azure-native/issues/2467?

**Comment by thomas11** [comment]
Just to check my understanding of what you're trying to do. Is it deploying essentially [this ARM template](https://learn.microsoft.com/en-us/azure/templates/github.network/networksettings?pivots=deployment-language-arm-template) but with Pulumi, using [Resource](https://www.pulumi.com/registry/packages/azure-native/api-docs/resources/resource)? But there's no way to specify the API version for github.network?

If so, I think #2467 will not help because it's about the API version sent to Azure, which is not the problem here.

Have you considered using the [Deployment](https://www.pulumi.com/registry/packages/azure-native/api-docs/resources/deployment/) resource, or its variants DeploymentAtSubscriptionScope etc., to deploy an ARM template for the github.network resource? Not very elegant but maybe a workaround.

**Comment by PhilBroderickCrezco** [comment]
> Just to check my understanding of what you're trying to do. Is it deploying essentially [this ARM template](https://learn.microsoft.com/en-us/azure/templates/github.network/networksettings?pivots=deployment-language-arm-template) but with Pulumi, using [Resource](https://www.pulumi.com/registry/packages/azure-native/api-docs/resources/resource)? But there's no way to specify the API version for github.network?
> 
> If so, I think [#2467](https://github.com/pulumi/pulumi-azure-native/issues/2467) will not help because it's about the API version sent to Azure, which is not the problem here.
> 
> Have you considered using the [Deployment](https://www.pulumi.com/registry/packages/azure-native/api-docs/resources/deployment/) resource, or its variants DeploymentAtSubscriptionScope etc., to deploy an ARM template for the github.network resource? Not very elegant but maybe a workaround.

Yes that's pretty much the gist of it. I'd like to avoid using ARM templates but it is an option. I'm also considering using the [Command resource](https://www.pulumi.com/registry/packages/command/) which we use for some other Azure workarounds. But again, these are workarounds and not very elegant as you mentioned.

**Comment by thomas11** [comment]
It would be great if one of you could share a failing Pulumi program that reproduces the issue. I checked the Azure API spec and the Pulumi resource seems to be _mostly_ correct based on that, but I haven't tried using it.

"Mostly" because the handling of the API version may be wrong, obviously. I think I was wrong above and that could be unblocked by #2467, or it could be handled by hard-coding an exception just for this resource to expose api-version as a resource property.

**Comment by PhilBroderickCrezco** [comment]
Any usage of `GitHub.Network/networkSettings` with the `Resource` type will fail, so something like this:

```csharp
 var resource = new Resource("github-network-settings-resource", new()
 {
      ResourceGroupName = "resourceGroupName"
      ParentResourcePath = "resourceGroupId",
      ResourceType = "GitHub.Network/networkSettings",
      ResourceProviderNamespace = "GitHub.Network",
      Location = "resourceGroupLocation",
      Properties = new Dictionary<string, object>
      {
           { "subnetId", "subnetId" },
           { "businessId", "businessId"}
      }
});
```
I tried with pinned versions of the `AzureNative` package but no luck. Also worth noting some properties (namely `ResourceProviderNamespace`, although not mentioned in the GitHub/Azure docs, is required.

I also attempted to include the version in the `ResourceType`, like in Terraform, so `GitHub.Network/networkSettings@2024-04-02` but as I expected that had no effect.

**Comment by thomas11** [comment]
I think we have the same issue with the `authorization.ManagementLock*` resources.

```
    + azure-native:authorization:ManagementLockAtResourceLevel: (create)
    ...
        resourceProviderNamespace: "Microsoft.Storage"
        resourceType             : "storageAccounts"

error: Code="InvalidResourceType" Message="The resource type could not be found in the namespace 'Microsoft.Storage' for api version '2020-05-01'."
```

**Comment by thomas11** [comment]
I tracked down the requests `az` and Pulumi make for the `GitHub.Network/networkSettings` resource.

Pulumi: https://management.azure.com/subscriptions/0282681f-7a9e-424b-80b2-96babd57a8a1/resourcegroups/resourceGroup6105d608/providers/GitHub
.Network/%2Fsubscriptions%2F0282681f-7a9e-424b-80b2-96babd57a8a1%2FresourceGroups%2FresourceGroup6105d608/GitHub.Network%2FnetworkSettings/github-network-settings-resource0b7f280d?api-version=2022-09-01

az: https://management.azure.com/subscriptions/0282681f-7a9e-424b-80b2-96babd57a8a1/resourcegroups/resourceGroupcc8d8bb0/providers/GitHub.Network//networkSettings/settingsFoo?api-version=2024-04-02

So there are two differences: the API version and the resource id.
